pramen/core/src/main/scala/za/co/absa/pramen/core/runner/task/TaskRunnerMultithreaded.scala

```diff
@@ -38,22 +39,57 @@ class TaskRunnerMultithreaded(conf: Config,
                               journal: Journal,
                               pipelineState: PipelineState,
                               runtimeConfig: RuntimeConfig) extends TaskRunnerBase(conf, bookkeeper, journal, runtimeConfig, pipelineState) {
+  private val log = LoggerFactory.getLogger(this.getClass)
+
   private val executor: ExecutorService = newFixedThreadPool(runtimeConfig.parallelTasks)
   implicit private val executionContext: ExecutionContextExecutorService = fromExecutorService(executor)
 
+  private val maxResources = runtimeConfig.parallelTasks
+  private val availableResources: Semaphore = new Semaphore(maxResources)
+
+  /**
+    * We want to prevent a case when a lot of resource-intensive actions are running at the same time.
+    * We have a finite number of resources. Before executing, the *action* asks for the required resources and only when
+    * there is enough of them available, the action will be executed.
+    */
+  private[task] def whenEnoughResourcesAreAvailable[T](requestedCount: Int)(action: => T): T = {
+    val resourceCount = getTruncatedResourceCount(requestedCount)
+
+    availableResources.acquire(resourceCount)
+    val result = Try { action }
+    availableResources.release(resourceCount)
+
+    result.get
```

Comment:
I like the idea of wrapping the dangerous code with `Try`, and using Try.get at the end!